packages/terra-disclosure-manager/src/DisclosureManager.jsx

```diff
@@ -0,0 +1,351 @@
+    newState.disclosureComponentKeys = Object.assign([], newState.disclosureComponentKeys);
+    newState.disclosureComponentData = Object.assign({}, newState.disclosureComponentData);
+
+    return newState;
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.resolveDismissPromise = this.resolveDismissPromise.bind(this);
+    this.resolveDismissChecksInSequence = this.resolveDismissChecksInSequence.bind(this);
+
+    this.safelyCloseDisclosure = this.safelyCloseDisclosure.bind(this);
+    this.generatePopFunction = this.generatePopFunction.bind(this);
+
+    this.renderContentComponents = this.renderContentComponents.bind(this);
+    this.renderDisclosureComponents = this.renderDisclosureComponents.bind(this);
+
+    this.openDisclosure = this.openDisclosure.bind(this);
+    this.pushDisclosure = this.pushDisclosure.bind(this);
+    this.popDisclosure = this.popDisclosure.bind(this);
+    this.closeDisclosure = this.closeDisclosure.bind(this);
+    this.requestDisclosureFocus = this.requestDisclosureFocus.bind(this);
+    this.releaseDisclosureFocus = this.releaseDisclosureFocus.bind(this);
+    this.maximizeDisclosure = this.maximizeDisclosure.bind(this);
+    this.minimizeDisclosure = this.minimizeDisclosure.bind(this);
+
+    // These cached functions are stored outside of state to prevent unnecessary rerenders.
+    this.dismissChecks = {};
+    this.onDismissResolvers = {};
+
+    this.state = {
+      disclosureIsOpen: false,
+      disclosureIsFocused: true,
+      disclosureIsMaximized: false,
+      disclosureSize: 'small',
+      disclosureComponentKeys: [],
+      disclosureComponentData: {},
+    };
+  }
+
+  openDisclosure(data) {
+    this.setState({
+      disclosureIsOpen: true,
+      disclosureIsFocused: true,
+      disclosureSize: data.size || availableDisclosureSizes.SMALL,
+      disclosureComponentKeys: [data.content.key],
+      disclosureComponentData: {
+        [data.content.key]: {
+          key: data.content.key,
+          name: data.content.name,
+          props: data.content.props,
+          component: data.content.component,
+        },
+      },
+    });
+  }
+
+  pushDisclosure(data) {
+    const newState = DisclosureManager.cloneDisclosureState(this.state);
+
+    newState.disclosureComponentKeys.push(data.content.key);
+    newState.disclosureComponentData[data.content.key] = {
+      key: data.content.key,
+      name: data.content.name,
+      props: data.content.props,
+      component: data.content.component,
+    };
+
+    this.setState(newState);
+  }
+
+  popDisclosure() {
+    const newState = DisclosureManager.cloneDisclosureState(this.state);
+
+    newState.disclosureComponentData[newState.disclosureComponentKeys.pop()] = undefined;
+
+    this.setState(newState);
+  }
+
+  closeDisclosure() {
+    this.setState({
+      disclosureIsOpen: false,
+      disclosureIsFocused: false,
+      disclosureIsMaximized: false,
+      disclosureSize: availableDisclosureSizes.SMALL,
```

Comment:
this seems awkward, should there be an undefined state?

Reply:
Maybe have a DEFAULT one that is 'small'?